cli/loader_win_utils.c

```diff
@@ -0,0 +1,198 @@
+// Workarounds for compiling via mingw without using libgcc_s
+typedef struct {
+    HANDLE fd;
+    BOOL isconsole;
+} FILE;
+
+static FILE _stdout = { INVALID_HANDLE_VALUE };
+static FILE _stderr = { INVALID_HANDLE_VALUE };
+
+FILE *stdout = &_stdout;
+FILE *stderr = &_stderr;
+
+int loader_fwrite(const WCHAR *str, size_t nchars, FILE *out) {
+    DWORD written;
+    if (out->isconsole) {
+        if (WriteConsole(out->fd, str, nchars, &written, NULL))
+            return written;
+    } else {
+        if (WriteFile(out->fd, str, sizeof(WCHAR) * nchars, &written, NULL))
+            return written;
+    }
+    return -1;
+}
+
+int loader_fputs(const char *str, FILE *out) {
+    wchar_t wstr[1024];
+    utf8_to_wchar(str, wstr, 1024);
+    return fwrite(wstr, wcslen(wstr), out);
+}
+
+void * loader_malloc(const size_t size) {
+    return HeapAlloc(GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, size);
+}
+
+void * loader_realloc(void * mem, const size_t size) {
+    return HeapReAlloc(GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, mem, size);
+}
+
+LPWSTR *CommandLineToArgv(LPWSTR lpCmdLine, int *pNumArgs) {
```

Comment:
Excuse my ignorance but is this copied from `shellapi.h`, curious why we can't just call that instead?

Reply:
We have to reimplement almost everything because we're writing a freestanding no-standard-libraries executable.  So although we can use definitions in header files, we can't link against almost any libraries, because most libraries will require things like a C support library (`libgcc_s`) and we don't want to link against that, because this loader needs to run _before_ we load any of our support libraries.